content/300-guides/150-testing/100-unit-testing.mdx

```diff
@@ -0,0 +1,311 @@
+
+```ts
+import { MockContext, Context, createMockContext } from '../context'
+
+let mockCtx: MockContext
+let ctx: Context
+
+beforeEach(() => {
+  mockCtx = createMockContext()
+  ctx = (mockCtx as unknown) as Context
+})
+```
+
+This will create a new context before each test is run via the `createMockContext` function. This (`mockCtx`) context will be used to make a mock call to Prisma and run a query to test. The `ctx` context will be used to run a scenario query that is tested against.
+
+## Example unit tests
+
+Prisma already tests query functions like `findMany`. A real world use case for unit testing Prisma might be a signup form. Your user fills in a form which calls a function, which in turn uses Prisma to make a call to your database.
+
+All of the examples that follow use the following schema model:
+
+```prisma file=schema.prisma
+model User {
+  id    Int     @id @default(autoincrement())
+  email String  @unique
+  name  String?
+}
+```
+
+The following unit tests will mock the process of
+
+- Creating a new user
+- Updating a users name
+
+The functions that use the dependency injection pattern will have the context injected (passed in as a parameter) into them, whereas the functions that use the singleton pattern will use the singleton instance of the Prisma client.
+
+<TabbedContent tabs={[<FileWithIcon text="functions-with-context.ts" icon="file"/>, <FileWithIcon text="functions-without-context.ts" icon="file"/>]}>
+
+<tab>
+
+```ts
+import { Context } from './context'
+
+interface CreateUser {
+  name: string
+  email: string
+}
+
+export async function createUser(user: CreateUser, ctx: Context) {
+  return await ctx.user.create({
+    data: user,
+  })
+}
+
+interface UpdateUser {
+  id: number
+  name: string
+  email: string
+}
+
+export async function updateUsername(user: UpdateUser, ctx: Context) {
+  return await ctx.user.update({
+    where: { id: user.id },
+    data: user,
+  })
+}
+```
+
+</tab>
+
+<tab>
+
+```ts
+import prisma from './client'
+
+interface CreateUser {
+  name: string
+  email: string
+}
+
+export async function createUser(user: CreateUser) {
+  return await prisma.user.create({
+    data: user,
+  })
+}
+
+interface UpdateUser {
+  id: number
+  name: string
+  email: string
+}
+
+export async function updateUsername(user: UpdateUser) {
+  return await prisma.user.update({
+    where: { id: user.id },
+    data: user,
+  })
+}
+```
+
+</tab>
+
+</TabbedContent>
+
+The tests for each methodology are fairly similar, the difference is how the mocked Prisma client is used.
+
+The **_dependency injection_** example passes the context through to the function that is being tested as well as using it to call the mock implementation.
+
+The **_singleton_** example uses the singleton client instance to call the mock implementation.
+
+<TabbedContent tabs={[<FileWithIcon text="with-singleton.ts" icon="file"/>, <FileWithIcon text="with-dependency-injection" icon="file"/>]}>
+
+<tab>
+
+```ts
```

Comment:
I wonder if we can extend these examples to test some _function-level logic_ rather than database logic. Right now it looks like we are testing that `user.update` and `user.create` work, which will always succeed. Why don't we test if the user has checked the `acceptTermsAndConditions` checkbox (should create user) vs if not (should show error message)? that's application-level logic rather than database-level.